cloud-engineering-with-gcp/python/content/lab-02/README.md

```diff
@@ -0,0 +1,199 @@
+
+Add the following statement near the top of your `__main__.py` near your other imports:
+
+```python
+import os
+```
+
+```python
+for file in ["404.html", "index.html"]:
+    filepath = os.path.join("../wwwroot", file)
+    gcp.storage.BucketObject(
+        file,
+        bucket=bucket.name,
+        name=file,
+        source=pulumi.FileAsset(filepath),
+        opts=pulumi.ResourceOptions(depends_on=[acl])
+    )
+```
+
+Notice the use of `depends_on`. This tells Pulumi that our `BucketObjects` should not be created until our `DefaultObjectAccessControl` has been fully provisioned. The `depends_on` is necessary because there's no _explicit_ dependency between the files and the ACL (i.e., there's no output from the `DefaultObjectAccessControl` resource passed to the `BucketObject` inputs). If we didn't explicitly specify `depends_on`, our files may get uploaded before the default ACL is applied, and our files would not get created with the right default permissions.
+
+## Step 4 &mdash; Run `pulumi up`
+
+Now that we've defined our infrastructure, we use the Pulumi CLI to create the resources we've defined.
+
+Run `pulumi up` in your project directory. You should see something like this:
+
+```bash
+$ pulumi up
+Previewing update (dev)
+
+View Live: https://app.pulumi.com/jaxxstorm/static_site/dev/previews/b5ce861f-16ac-4930-ada4-1e9547e02eb1
+
+     Type                                       Name             Plan
+ +   pulumi:pulumi:Stack                        static_site-dev  create
+ +   ├─ gcp:storage:Bucket                      website          create
+ +   ├─ gcp:storage:DefaultObjectAccessControl  website          create
+ +   ├─ gcp:storage:BucketObject                404.html         create
+ +   └─ gcp:storage:BucketObject                index.html       create
+
+Resources:
+    + 5 to create
+
+Do you want to perform this update?  [Use arrows to move, enter to select, type to filter]
+  yes
+> no
+```
+
+You can examine the details of the resources that will be created. When you're happy, move the arrow to `yes` and watch as Pulumi creates your resources!
+
+## Step 5 &mdash; Export the Bucket URL
+
+Our final step is to build our static site URL and add it as a [stack output](https://www.pulumi.com/learn/building-with-pulumi/stack-outputs/).
+
+First, we assemble the output's value using `pulumi.Output.concat`:
+
+```python
+static_site_url = pulumi.Output.concat(
+    "https://storage.googleapis.com/", bucket.name, "/index.html")
+```
+
+We use `pulumi.Output.concat` instead of standard Python string concatenation because `bucket.name` is a Pulumi output - a value that isn't known until after a resource has been created. For more information on Pulumi inputs and outputs, see [Inputs and Outputs](https://www.pulumi.com/docs/intro/concepts/inputs-outputs/) in the Pulumi docs.
```

Comment:
```suggestion
We use `pulumi.Output.concat` instead of standard Python string concatenation because `bucket.name` is a Pulumi Output&mdash;a value that isn't known until after a resource has been created. For more information on Pulumi Inputs and Outputs, reference [Inputs and Outputs](https://www.pulumi.com/docs/intro/concepts/inputs-outputs/) in the Pulumi docs.
```